### PR TITLE
[MPS] Add prelu

### PR DESCRIPTION
**Pull request by qqaatw**
### Description
Adds `prelu` MPS support.

I'm wondering whether the generic test framework implemented in `test_mps.py` checks the correctness of gradients.

### CODE REVIEW

**Comment by kulinseth** [comment]
> I'm wondering whether the generic test framework implemented in test_mps.py checks the correctness of gradients.

I don't think it checks the gradient ops. @albanD to confirm.

File: aten/src/ATen/native/mps/operations/Activation.mm [hunk 1]
```diff
@@ -1633,6 +1633,253 @@ Tensor glu_backward_mps (const Tensor& grad_output,
...
+      if (ndim >= 2) {
+        sizes[1] = t.dim() == 1 ? t.size(0) : 1;
+        strides[1] = t.dim() == 1 ? t.stride(0) : 0;
+        return t.as_strided(sizes, strides);
+      }
+      return t.as_strided(sizes, strides);
+    };
+    Tensor w = as_nd(weight_);
+
+      struct CachedGraph : public MPSCachedGraph
```
Comment (kulinseth) [inline aten/src/ATen/native/mps/operations/Activation.mm:1678]: Indentation.

File: aten/src/ATen/native/mps/operations/Activation.mm [hunk 2]
```diff
@@ -1633,6 +1633,253 @@ Tensor glu_backward_mps (const Tensor& grad_output,
...
+  const int64_t ndim = self.dim();
+  // Helper to convert 1d tensor or scalar tensor to an nd tensor that broadcasts with input
+  // All elements go into the channel dimension
+  DimVector sizes(ndim, 1), strides(ndim, 0);
+  auto as_nd = [&](const Tensor& t) {
+    TORCH_INTERNAL_ASSERT(t.defined() && (t.dim() == 1 || t.dim() == 0));
+    if (ndim >= 2) {
+      sizes[1] = t.dim() == 1 ? t.sizes()[0] : 1;
+      strides[1] = t.dim() == 1 ? t.strides()[0] : 0;
+      return t.as_strided(sizes, strides);
```
Comment (kulinseth) [inline aten/src/ATen/native/mps/operations/Activation.mm:1779]: Same comment as above.

File: aten/src/ATen/native/mps/operations/Activation.mm [hunk 3]
```diff
@@ -1633,6 +1633,253 @@ Tensor glu_backward_mps (const Tensor& grad_output,
...
+
+  NSMutableArray<NSNumber*> * reduce_dims = [NSMutableArray<NSNumber*> new];
+  int64_t input_ndim = self.dim();
+
+  for (const auto i : c10::irange(input_ndim)) {
+    if (weight_num != 1 && i == 1) continue;
+    [reduce_dims addObject:[NSNumber numberWithInt:i]];
+  }
+
+      struct CachedGraph : public MPSCachedGraph
```
Comment (kulinseth) [inline aten/src/ATen/native/mps/operations/Activation.mm:1793]: Indentation.

**Review by kulinseth — COMMENTED** [review]
Overall the change looks good, thanks @qqaatw. If you feel that separating out the graphs for t.dim()==1 is too much work, please add a TODO: "current implementation can be improved by separating out the t.dim==1 case by avoiding the use of as_strided."

File: aten/src/ATen/native/mps/operations/Activation.mm [hunk 4]
```diff
@@ -1633,6 +1633,253 @@ Tensor glu_backward_mps (const Tensor& grad_output,
...
+    auto as_nd = [&](const Tensor& t) {
+      TORCH_CHECK(
+        t.dim() == 1 || t.dim() == 0,
+        "prelu: Expected `weight` to be a scalar or 1D tensor, but got ndim = ", t.dim());
+      if (ndim >= 2) {
+        sizes[1] = t.dim() == 1 ? t.size(0) : 1;
+        strides[1] = t.dim() == 1 ? t.stride(0) : 0;
+        return t.as_strided(sizes, strides);
+      }
+      return t.as_strided(sizes, strides);
```
[the remark this answers is not in the document]
Comment (qqaatw) [reply aten/src/ATen/native/mps/operations/Activation.mm]: Done.

**Comment by albanD** [comment]
> I don't think it checks the gradient ops. @albanD to confirm.

It does not, but we can add that relatively easily: add another very similar test that request sample inputs that require grad, run backward and compare the computed gradients!

**Review by kulinseth — APPROVED** [review]
Looks great, thanks.

**Pull request closed without merging**